### ISSUE TITLE
Is there a way to update a config of local SF cluster fully if certificate is expired?

### ISSUE THREAD

**Issue by strymm** [Azure/service-fabric-issues#1560]
We have a local service fabric cluster running on several internal machines (NOT in Azure Cloud) and use a certificate to secure it.

Usually we used following way to update a certificate (if it's not expired yet):

1. Create new self-signed certificate using CertSetup.ps1 from service fabric SDK
2. Install it on all nodes.
3. `Get-ServiceFabricClusterConfiguration > C:\SfInstall\ClusterConfig.current.json`
4. Edit ClusterConfig.current.json to replace old thumbprint to new one
5. `Start-ServiceFabricClusterConfigurationUpgrade -ClusterConfigPath C:\SfInstall\ClusterConfig.current.json`

It worked fine and updated certificate fully (I mean here the cluster worked fine and commands `Get-ServiceFabricClusterConfiguration` and `Get-ServiceFabricClusterManifest` showed the new thumbprint after that)

But some time ago our certificate expired and we had to follow the instruction
https://github.com/Azure/Service-Fabric-Troubleshooting-Guides/blob/3754d5c2d4bfe6094bff000cb7ffd4f80897be31/Security/Fix%20Expired%20Cluster%20Certificate%20Manual%20Steps.md
to update certificate and get the cluster worked.

It allowed to get the cluster worked but commands `Get-ServiceFabricClusterConfiguration` and `Get-ServiceFabricClusterManifest` still showed old thumbprint (seems it's due to note 2 from point 10 in the instruction above).

We managed to update a cluster manifest following steps:

1. Run `Get-ServiceFabricRegisteredClusterConfigVersion` to detect current version of the cluster manifest
2. `Get-ServiceFabricClusterManifest > C:\SfInstall\ClusterManifestV7.xml`
3. Edit ClusterManifestV7.xml to replace old thumbprint to new one
4. `Copy-ServiceFabricClusterPackage -ImageStoreConnectionString "fabric:ImageStore" -Config -ClusterManifestPath "C:\SfInstall\ClusterManifestV7.xml"`
5. `Register-ServiceFabricClusterPackage -Config -ClusterManifestPath "ClusterManifestV7.xml"`
6. `Start-ServiceFabricClusterUpgrade -Monitored -FailureAction Rollback -Config -ClusterManifestVersion 7`

After that `Get-ServiceFabricClusterManifest` became to show new thumbprint
![Get-ServiceFabricClusterManifest](https://user-images.githubusercontent.com/53749800/62453869-eefe1480-b77b-11e9-8c5d-c2899f1fd52a.png)

but `Get-ServiceFabricClusterConfiguration` still shows old one
![Get-ServiceFabricClusterConfiguration](https://user-images.githubusercontent.com/53749800/62454062-4603e980-b77c-11e9-8604-378783683c2e.png)

and isn't possible to update cluster config in usual way using steps

3. `Get-ServiceFabricClusterConfiguration > C:\SfInstall\ClusterConfig.current.json`
4. Edit ClusterConfig.current.json to replace old thumbprint to new one
5. `Start-ServiceFabricClusterConfigurationUpgrade -ClusterConfigPath C:\SfInstall\ClusterConfig.current.json`

due to error
![Start-ServiceFabricClusterConfigurationUpgrade](https://user-images.githubusercontent.com/53749800/62454092-56b45f80-b77c-11e9-95db-ae0fedc9ffe6.png)

So, my questions are
1. What is the difference between upgrade of cluster config using `Start-ServiceFabricClusterUpgrade` with flag `-Config` and `Start-ServiceFabricClusterConfigurationUpgrade`? Why first approach works but second one gets failed in case of updating expired certificate?
2. Is there a way (or workaround) to update a config of local SF cluster fully in order to `Get-ServiceFabricClusterConfiguration` to show new thumbprint and become possible to update cluster config in usual way using `Start-ServiceFabricClusterConfigurationUpgrade`?

**Comment by masnider** [comment]
@dkkapur and @dragav 

The steps should be similar to this: https://github.com/Azure/Service-Fabric-Troubleshooting-Guides/blob/master/Security/Fix%20Expired%20Cluster%20Certificate%20Manual%20Steps.md

**Comment by strymm** [comment]
@masnider
Yeah, we also thought in this way but in fact seems there are some differences for on-premise SF standalone cluster.

We performed steps 7-14 from the instruction
https://github.com/Azure/Service-Fabric-Troubleshooting-Guides/blob/master/Security/Fix%20Expired%20Cluster%20Certificate%20Manual%20Steps.md
and after that we recovered SF cluster and command
`Start-ServiceFabricClusterUpgrade`
became to work fine but command
`Start-ServiceFabricClusterConfigurationUpgrade`
still throws error mentioned above.

We even updated SF runtime to 7.0.457.9590 and cluster manifest (using `Start-ServiceFabricClusterUpgrade`) to turn on `AcceptExpiredPinnedClusterCertificate` setting as mentioned at
https://github.com/Azure/Service-Fabric-Troubleshooting-Guides/blob/master/Security/How%20to%20recover%20from%20an%20Expired%20Cluster%20Certificate.md
but it still didn't help.

So, main question is **how to upgrade Service Fabric Resource Provider (SFRP) records with the new thumbprint for ON-PREMISE SF standalone cluster** (see note 2 from point 14 in the instruction provided you above)? Because seems the problem is here.
![SfResourceProvider](https://user-images.githubusercontent.com/53749800/71324332-b1154900-24ee-11ea-997a-b43f8ed7bc86.png)

**Comment by dragav** [comment]
@strymm Start-ServiceFabricClusterUpgrade follows a different code path than Start-ServiceFabricClusterConfigurationUpgrade; the former takes in the cluster manifest in xml format and is 'closer' to the actual flow that would be invoked when upgrading an Azure cluster via SFRP. The latter takes in a json template and runs some pre-checks in the process of converting it to a cluster manifest.
It seems that we have a bug in that second code path, which we'll need to fix.
